modules/internal/localeModels/apu/LocaleModel.chpl

```diff
@@ -0,0 +1,295 @@
+    }
+
+
+    proc GPULocale(_sid, _parent) {
+      sid = _sid;
+      parent = _parent;
+      name = parent.chpl_name() + "-GPU" + sid;
+    }
+
+    proc writeThis(f) {
+      parent.writeThis(f);
+      f <~> '.'+name;
+    }
+
+    proc getChildCount(): int { return 0; }
+    iter getChildIndices() : int {
+      halt("No children to iterate over.");
+      yield -1;
+    }
+    proc addChild(loc:locale) {
+      halt("Cannot add children to this locale type.");
+    }
+    proc getChild(idx:int) : locale { return nil; }
+  }
+
+  const chpl_emptyLocaleSpace: domain(1) = {1..0};
+  const chpl_emptyLocales: [chpl_emptyLocaleSpace] locale;
+
+  //
+  // The node model
+  //
+  class LocaleModel : AbstractLocaleModel {
+    const _node_id : int;
+    const local_name : string;
+
+    const numSublocales: int;
+    var GPU: GPULocale;
+    var CPU: CPULocale;
+
+    // This constructor must be invoked "on" the node
+    // that it is intended to represent.  This trick is used
+    // to establish the equivalence the "locale" field of the locale object
+    // and the node ID portion of any wide pointer referring to it.
+    proc LocaleModel() {
+      if doneCreatingLocales {
+        halt("Cannot create additional LocaleModel instances");
+      }
+      setup();
+    }
+
+    proc LocaleModel(parent_loc : locale) {
+      if doneCreatingLocales {
+        halt("Cannot create additional LocaleModel instances");
+      }
+      parent = parent_loc;
+      setup();
+    }
+
+    proc chpl_id() return _node_id;     // top-level locale (node) number
+    proc chpl_localeid() {
+      return chpl_buildLocaleID(_node_id:chpl_nodeID_t, c_sublocid_any);
+    }
+    proc chpl_name() return local_name;
+
+
+    proc writeThis(f) {
+      // Most classes will define it like this:
+      //      f <~> name;
+      // but here it is defined thus for backward compatibility.
+      f <~> new ioLiteral("LOCALE") <~> _node_id;
+    }
+
+    proc getChildSpace() {
+      halt("error!");
+      return {0..#numSublocales};
+    }
+
+    proc getChildCount() return 0;
+
+    iter getChildIndices() : int {
+      for idx in {0..#numSublocales} do // chpl_emptyLocaleSpace do
+        yield idx;
+    }
+
+    proc getChild(idx:int) : locale {
+      if idx == 1
+        then return GPU;
+      else
+        return CPU;
+    }
+
+    iter getChildren() : locale  {
+      halt("in here");
+      for idx in {0..#numSublocales} {
+        if idx == 1
+          then yield GPU;
+        else
+          yield CPU;
+      }
+    }
+
+    proc getChildArray() {
+      halt ("in get child Array");
+      return chpl_emptyLocales;
+    }
+
+
+    //------------------------------------------------------------------------{
+    //- Implementation (private)
+    //-
+    proc setup() {
+      _node_id = chpl_nodeID: int;
+
+      helpSetupLocaleAPU(this, local_name, numSublocales);
+    }
+    //------------------------------------------------------------------------}
+ }
```

Comment:
Oh, great, I'm glad this got added. I think this was actually a change that we requested because we needed a place to clean up the HSA runtime. I'll probably add a commented out cleanup routine in the deinit()